tools/codegen/api/types.py

```diff
@@ -40,15 +51,179 @@ def __str__(self) -> str:
 
+# Describe an implicit this argument (*this) on methods in the C++ API
+@dataclass(frozen=True)
+class CppThisArgument:
+    # The grouped JIT argument this formal was derived from
+    argument: ThisArgument
+
+    # C++ type, e.g., Tensor&
+    type: str
+
+    # this arguments are never defaulted
+    def no_default(self) -> 'CppThisArgument':
+        return self
+
+    # The this argument is implicit, so it's not included in the
+    # explicit arguments list.
+    def explicit_arguments(self) -> Sequence[CppArgument]:
+        return []
+
+# Semantically represents a bundle of CppArguments that collectively
+# represent a TensorOptions.  If you don't care about TensorOptions
+# processing, think of this as just a list of four CppArguments; however
+# if you need to bundle these arguments back into a single
+# TensorOptions, it will be easiest to operate on this struct as a
+# whole.
+@dataclass(frozen=True)
+class CppTensorOptionsArguments:
+    argument: TensorOptionsArguments
+    dtype: CppArgument
+    layout: CppArgument
+    device: CppArgument
+    pin_memory: CppArgument
+
+    # Remove the defaults from each of the constituent arguments
+    # representing the TensorOptions
+    def no_default(self) -> 'CppTensorOptionsArguments':
+        return CppTensorOptionsArguments(
+            argument=self.argument,
+            dtype=self.dtype.no_default(),
+            layout=self.layout.no_default(),
+            device=self.device.no_default(),
+            pin_memory=self.pin_memory.no_default(),
+        )
+
+    # Flatten the TensorOptions into individual CppArguments
+    def explicit_arguments(self) -> Sequence[CppArgument]:
+        return [self.dtype, self.layout, self.device, self.pin_memory]
+
 @dataclass(frozen=True)
 class CppExpr:
     type: str
     expr: str
 
+# A CppSignature represents a single overload in the C++ API.  For
+# any given function schema, there may be multiple CppSignatures
+# corresponding to it, based on how we desugar to C++.  See also
+# CppSignatureGroup.
+@dataclass(frozen=True)
+class CppSignature:
+    # The schema this signature is derived from
+    func: FunctionSchema
+
+    # Enough information about the C++ types to generate a full
+    # C++ type signature for this signature.  I'm not too sure
+    # if these are the right representations, so for now this
+    # is intended to be more abstract.  Prefer using
+    # cpp_grouped_arguments() to access this
+    _cpp_grouped_arguments: Tuple[Union[CppArgument, CppTensorOptionsArguments, CppThisArgument], ...]
+    _cpp_returns_type: str
+
+    # WARNING: This is probably NOT what you want
+    #
+    # Return the flattened argument structure of this signature,
+    # discarding information about grouping.  If you are planning
+    # to translate these arguments into expressions on another
+    # API, you probably want cpp_grouped_arguments instead (which
+    # will handle TensorOptions translations correctly)
```

Comment:
I would add a comment here that this removes the implicit this argument. Maybe we should also come up with a better name, since "grouped" vs "ungrouped" doesn't catch the this argument handling, but I don't have an idea right now. It's basically pre-transformation and post-transformation arguments, but that doesn't help with naming. Note that in the same spirit, the CppArgument type returned from `cpp_ungrouped_arguments` should probably be a different type than the CppArgument returned from `cpp_grouped_arguments`, since the latter are pre-transformation.